app/Foliage/PreparePackageVersion.hs

```diff
@@ -39,6 +40,7 @@ data PreparedPackageVersion = PreparedPackageVersion
     pkgTimestamp :: Maybe UTCTime,
     pkgVersionSource :: PackageVersionSource,
     pkgVersionForce :: Bool,
+    pkgVersionDeprecated :: Bool,
```

Comment:
It's important that we keep all the deprecation changes here, because each of them will give a separate contribution to the index. The final deprecation status can be precomputed in addition to that. You can see this is similar to what we do with revisions (i.e. `cabalFileRevisions` and `cabalFilePath`).

```suggestion
    pkgVersionIsDeprecated :: Bool,
    pkgVersionDeprecationChanges :: [(UTCTime, Bool)]
```

Reply:
Btw, it would be nice if we could guarantee that these lists `[(UTCTime, a)]` are sorted. Maybe using `sorted-list`? Idk.